File: src/solutions2024.rs
use crate::Answer;

use std::collections::HashMap;

pub fn solution_2024_01(input: &str) -> Answer {
    let mut answer = Answer {
        first: Some(0),
        second: Some(0),
    };
    let mut list1: Vec<i64> = Vec::new();
    let mut list2: Vec<i64> = Vec::new();
    let mut hash: HashMap<i64, i64> = HashMap::new();

    for line in input.lines() {
        let values: Vec<i64> = line
            .split_whitespace()
            .filter_map(|number| number.parse::<i64>().ok())
            .collect();

        if values.len() == 2 {
            list1.push(values[0]);
            list2.push(values[1]);
            *hash.entry(values[1]).or_insert(0) += 1;
        }
    }

    list1.sort();
    list2.sort();

    for (index, number) in list1.iter().enumerate() {
        let check: i64 = *hash.get(number).unwrap_or(&0);

        answer.first = Some(answer.first.unwrap_or(0) + (number - list2[index]).abs());
        answer.second = Some(answer.second.unwrap_or(0) + number * check);
    }

    answer
}

pub fn solution_2024_02(input: &str) -> Answer {
    let mut answer = Answer {
        first: Some(0),
        second: Some(0),
    };

    for line in input.lines() {
        let mut check: bool = false;
        let mut reports: Vec<Vec<i64>> = Vec::new();

        reports.push(
            line.split_whitespace()
                .filter_map(|number| number.parse::<i64>().ok())
                .collect(),
        );

        for index in 0..reports[0].len() {
            let mut copy = reports[0].clone();

            copy.remove(index);
            reports.push(copy);
        }

        for (index, report) in reports.iter().enumerate() {
            let mut errors: i64 = 0;

            if report.is_sorted() || report.is_sorted_by(|a, b| b <= a) {
                for window in report.windows(2) {
                    let difference: i64 = (window[1] - window[0]).abs();

                    if difference < 1 || difference > 3 {
                        errors += 1
                    }
                }
            } else {
                errors += 1;
            }

            match (index, errors) {
                (0, 0) => {
                    answer.first = Some(answer.first.unwrap_or(0) + 1);
                }
                (1.., 0) => check = true,
                _ => {}
            }
        }

        if check {
            answer.second = Some(answer.second.unwrap_or(0) + 1);
        }
    }

    answer
}

#[cfg(test)]
mod tests_2024 {
    use super::*;

    #[test]
    fn test_2024_01() {
        let input: &str = r#"3   4
4   3
2   5
1   3
3   9
3   3"#;

        assert_eq!(solution_2024_01(input).first, Some(11));
        assert_eq!(solution_2024_01(input).second, Some(31));
    }

    #[test]
    fn test_2024_02() {
        assert_eq!(solution_2024_02("7 6 4 2 1").first, Some(1));
        assert_eq!(solution_2024_02("1 2 7 8 9").first, Some(0));
        assert_eq!(solution_2024_02("9 7 6 2 1").first, Some(0));
        assert_eq!(solution_2024_02("1 3 2 4 5").first, Some(0));
        assert_eq!(solution_2024_02("1 3 6 7 9").first, Some(1));

        assert_eq!(solution_2024_02("7 6 4 2 1").second, Some(1));
        assert_eq!(solution_2024_02("1 2 7 8 9").second, Some(0));
        assert_eq!(solution_2024_02("9 7 6 2 1").second, Some(0));
        assert_eq!(solution_2024_02("1 3 2 4 5").second, Some(1));
        assert_eq!(solution_2024_02("1 3 6 7 9").second, Some(1));
    }
}

File: Sources/advent/Solutions2024.swift
import Foundation

func solution_2024_01(_ input: String) -> Answer {
    var answer = Answer()
    var list1: [Int] = []
    var list2: [Int] = []
    var hash: [Int: Int] = [:]

    for line in input.components(separatedBy: .newlines) where line.isEmpty == false {
        let values: [Int] = line.split(separator: " ")
            .compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
        
        if values.count == 2 {
            list1.append(values[0])
            list2.append(values[1])
            hash[values[1], default: 0] += 1
        }
    }
    
    list1.sort()
    list2.sort()
    
    for (index, number) in list1.enumerated() {
        let check = hash[number] ?? 0
        
        answer.first += abs(number - list2[index])
        answer.second += number * check
    }
    
    return answer
}

func solution_2024_02(_ input: String) -> Answer {
    var answer = Answer()
    
    for line in input.components(separatedBy: .newlines) where line.isEmpty == false {
        var check = false
        var reports: [[Int]] = []
        
        reports.append(
            line.split(separator: " ")
                .compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
        )
        
        for index in 0..<reports[0].count {
            var copy = reports[0]
            copy.remove(at: index)
            reports.append(copy)
        }
        
        for (index, report) in reports.enumerated() {
            var errors = 0
            
            let ascending = report.sorted()
            let descending = report.sorted(by: >)
            
            if report == ascending || report == descending {
                for i in 0..<report.count - 1 {
                    let difference = abs(report[i + 1] - report[i])
                    
                    if difference < 1 || difference > 3 {
                        errors += 1
                    }
                }
            } else {
                errors += 1
            }
            
            switch (index, errors) {
            case (0, 0):
                answer.first += 1
            case (1..., 0):
                check = true
            default:
                break
            }
        }
        
        if check {
            answer.second += 1
        }
    }
    
    return answer
}

